### PR TITLE
Add JSON output option to DockerInfo.

### PR DESCRIPTION
**Pull request by jonpspri**
As part of a multi-architecture build, I need to be able access variables from the DockerInfo of the Docker with which I am interacting.  This adds a parameter (dockerInfoFile) to the DockerInfo task which, if set, will trigger a write of JSON representing the DockerInfo received from the Docker instance.

### CODE REVIEW

**Comment by orzeh** [comment]
@jonpspri thanks for the PR! I think it would be better to follow the convention we use in other tasks with respect to handling results - let's use [`onNext`](https://github.com/bmuschko/gradle-docker-plugin#onnext) handler and pass there object received from Docker daemon. It will be more general solution and one can save the result as JSON or do whatever is needed.

**Comment by jonpspri** [comment]
Fair 'nuff.  Let me take a look at how the onNext handling works.  Stand by... :)

File: src/main/groovy/com/bmuschko/gradle/docker/tasks/DockerInfo.groovy [hunk 1]
```diff
@@ -37,5 +41,9 @@ class DockerInfo extends AbstractDockerRemoteApiTask {
         logger.quiet "NGoroutines          : $info.nGoroutines"
         logger.quiet "Swap Limit           : $info.swapLimit"
         logger.quiet "Execution Driver     : $info.executionDriver"
+
+        if (onNext) {
+            onNext.call(info)
+        }
```
Comment (orzeh) [inline src/main/groovy/com/bmuschko/gradle/docker/tasks/DockerInfo.groovy]: Can you add `else` here and move logger statements into it?

Comment (jonpspri) [reply src/main/groovy/com/bmuschko/gradle/docker/tasks/DockerInfo.groovy]: Oh... so display Or onNext but not both.  Hmm... I kind if liked still having the logger, but that might be a matter of personal taste.  I suppose one can always manually log from within the onNext.

**Comment by jonpspri** [comment]
Also... will start working on Functional test.  Thanks for the example pointer.

File: src/main/groovy/com/bmuschko/gradle/docker/tasks/DockerInfo.groovy (hunk 1 above)
Comment (jonpspri) [reply src/main/groovy/com/bmuschko/gradle/docker/tasks/DockerInfo.groovy]: Done.

**Comment by jonpspri** [comment]
Done.

File: src/functTest/groovy/com/bmuschko/gradle/docker/DockerInfoFunctionalTest.groovy [hunk 2]
```diff
@@ -17,4 +17,44 @@ class DockerInfoFunctionalTest extends AbstractFunctionalTest {
         then:
         result.output.contains('Retrieving Docker info.')
     }
+
+    def "Calls onNext when provided"() {
```
Comment (cdancy) [inline src/functTest/groovy/com/bmuschko/gradle/docker/DockerInfoFunctionalTest.groovy]: Can we can combine the expectations of this test with `Passes dockerinfo to onNext`?

**Comment by cdancy** [comment]
One minor comment to address as I think we can combine the functional tests written into 1. Other than that it LGTM.

**Comment by jonpspri** [comment]
Yeah, I was thinking about pulling them together after I wrote the two of them and saw how similar they were.  There are different schools of test cases -- "Write lots because the differences between them can sometimes add fidelity to diagnosing a problem" and "Write fewer and still cover cases because it's less to maintain."  I usually adhere to the former school, but given how heavy these integration cases are (after watching my Mac churn for an hour), I'm happy to move toward the latter.  Stand by for commit...

**Comment by jonpspri** [comment]
et voila.

**Comment by cdancy** [comment]
@jonpspri thanks for the PR!

**Pull request merged**